File: desktop_wallet/src/services/crypto_card_service.cpp
// Crypto Card Service Implementation
#include "crypto_card_service.h"
#include <sstream>
#include <iomanip>
#include <random>
#include <algorithm>

namespace tigerwallet {

CryptoCardService::CryptoCardService() {}
CryptoCardService::~CryptoCardService() {}

std::vector<CryptoCard> CryptoCardService::getCards(const std::string& userId) {
    if (cards_.find(userId) != cards_.end()) {
        return cards_[userId];
    }
    return std::vector<CryptoCard>();
}

std::string generateCardNumber() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 9);
    
    std::string number = "4532";
    for (int i = 0; i < 12; i++) {
        number += std::to_string(dis(gen));
    }
    return number;
}

std::string generateCVV() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 999);
    std::stringstream ss;
    ss << std::setw(3) << std::setfill('0') << dis(gen);
    return ss.str();
}

std::string generateExpiry() {
    time_t now = time(nullptr);
    tm* local = localtime(&now);
    int month = (local->tm_mon + 1) % 12 + 1;
    int year = (local->tm_year + 3) % 100;
    std::stringstream ss;
    ss << std::setw(2) << std::setfill('0') << month << "/" << year;
    return ss.str();
}

CryptoCard CryptoCardService::createVirtualCard(const std::string& userId, const std::string& cardHolder,
                                                 const std::string& type, const std::string& network,
                                                 double dailyLimit, double monthlyLimit) {
    CryptoCard card;
    std::stringstream ss;
    ss << "card_" << time(nullptr);
    card.id = ss.str();
    card.userId = userId;
    card.cardNumber = generateCardNumber();
    card.cardHolder = cardHolder;
    card.expiryDate = generateExpiry();
    card.cvv = generateCVV();
    card.type = type;
    card.network = network;
    card.status = "ACTIVE";
    card.dailyLimit = dailyLimit;
    card.monthlyLimit = monthlyLimit;
    card.dailySpent = 0;
    card.monthlySpent = 0;
    card.applePayEnabled = true;
    card.googlePayEnabled = true;
    card.createdAt = time(nullptr);
    
    cards_[userId].push_back(card);
    return card;
}

CryptoCard CryptoCardService::createPhysicalCard(const std::string& userId, const std::string& cardHolder,
                                                 const std::string& shippingAddress,
                                                 double dailyLimit, double monthlyLimit) {
    CryptoCard card = createVirtualCard(userId, cardHolder, "PHYSICAL", "VISA", dailyLimit, monthlyLimit);
    card.status = "PENDING_ACTIVATION";
    return card;
}

void CryptoCardService::freezeCard(const std::string& userId, const std::string& cardId) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.status = "FROZEN";
            break;
        }
    }
}

void CryptoCardService::unfreezeCard(const std::string& userId, const std::string& cardId) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.status = "ACTIVE";
            break;
        }
    }
}

void CryptoCardService::terminateCard(const std::string& userId, const std::string& cardId) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.status = "TERMINATED";
            card.applePayEnabled = false;
            card.googlePayEnabled = false;
            break;
        }
    }
}

std::vector<CardTransaction> CryptoCardService::getTransactions(const std::string& userId, const std::string& cardId) {
    if (transactions_.find(userId) != transactions_.end()) {
        if (!cardId.empty()) {
            std::vector<CardTransaction> result;
            for (const auto& txn : transactions_[userId]) {
                if (txn.cardId == cardId) {
                    result.push_back(txn);
                }
            }
            return result;
        }
        return transactions_[userId];
    }
    return std::vector<CardTransaction>();
}

CardTransaction CryptoCardService::processPayment(const std::string& userId, const std::string& cardId,
                                                   double amount, const std::string& currency,
                                                   const std::string& merchantName) {
    CardTransaction txn;
    std::stringstream ss;
    ss << "txn_" << time(nullptr);
    txn.id = ss.str();
    txn.cardId = cardId;
    txn.userId = userId;
    txn.amount = amount;
    txn.currency = currency;
    txn.merchantName = merchantName;
    txn.merchantCategory = "RETAIL";
    txn.type = "PURCHASE";
    txn.status = "COMPLETED";
    txn.timestamp = time(nullptr);
    
    // Update card spending
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.dailySpent += amount;
            card.monthlySpent += amount;
            break;
        }
    }
    
    transactions_[userId].push_back(txn);
    return txn;
}

std::vector<CardFundingSource> CryptoCardService::getFundingSources(const std::string& userId) {
    if (fundingSources_.find(userId) != fundingSources_.end()) {
        return fundingSources_[userId];
    }
    return std::vector<CardFundingSource>();
}

void CryptoCardService::addFundingSource(const std::string& userId, const std::string& token,
                                         const std::string& symbol, double balance, bool isDefault) {
    CardFundingSource source;
    std::stringstream ss;
    ss << "fund_" << time(nullptr);
    source.id = ss.str();
    source.userId = userId;
    source.token = token;
    source.symbol = symbol;
    source.balance = balance;
    source.dailyLimit = 10000;
    source.monthlyLimit = 100000;
    source.isDefault = isDefault;
    
    fundingSources_[userId].push_back(source);
}

void CryptoCardService::setCardLimit(const std::string& userId, const std::string& cardId,
                                     double dailyLimit, double monthlyLimit) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.dailyLimit = dailyLimit;
            card.monthlyLimit = monthlyLimit;
            break;
        }
    }
}

void CryptoCardService::enableApplePay(const std::string& userId, const std::string& cardId) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.applePayEnabled = true;
            break;
        }
    }
}

void CryptoCardService::enableGooglePay(const std::string& userId, const std::string& cardId) {
    auto& userCards = cards_[userId];
    for (auto& card : userCards) {
        if (card.id == cardId) {
            card.googlePayEnabled = true;
            break;
        }
    }
}

} // namespace tigerwallet

File: mobile/flutter/lib/features/crypto_card/crypto_card_service.dart
// Crypto Card Service - Flutter Implementation
// Virtual and Physical Crypto Cards with Apple Pay/Google Pay support

class CryptoCard {
  final String id;
  final String userId;
  final String cardNumber;
  final String cardHolder;
  final String expiryDate;
  final String cvv;
  final String type;
  final String network;
  final String status;
  final double dailyLimit;
  final double monthlyLimit;
  final double dailySpent;
  final double monthlySpent;
  final String? virtualCardImage;
  final bool applePayEnabled;
  final bool googlePayEnabled;
  final DateTime createdAt;

  CryptoCard({
    required this.id,
    required this.userId,
    required this.cardNumber,
    required this.cardHolder,
    required this.expiryDate,
    required this.cvv,
    required this.type,
    required this.network,
    required this.status,
    required this.dailyLimit,
    required this.monthlyLimit,
    required this.dailySpent,
    required this.monthlySpent,
    this.virtualCardImage,
    required this.applePayEnabled,
    required this.googlePayEnabled,
    required this.createdAt,
  });

  factory CryptoCard.fromJson(Map<String, dynamic> json) {
    return CryptoCard(
      id: json['id'] ?? '',
      userId: json['userId'] ?? '',
      cardNumber: json['cardNumber'] ?? '',
      cardHolder: json['cardHolder'] ?? '',
      expiryDate: json['expiryDate'] ?? '',
      cvv: json['cvv'] ?? '',
      type: json['type'] ?? 'VIRTUAL',
      network: json['network'] ?? 'VISA',
      status: json['status'] ?? 'ACTIVE',
      dailyLimit: (json['dailyLimit'] ?? 10000).toDouble(),
      monthlyLimit: (json['monthlyLimit'] ?? 100000).toDouble(),
      dailySpent: (json['dailySpent'] ?? 0).toDouble(),
      monthlySpent: (json['monthlySpent'] ?? 0).toDouble(),
      virtualCardImage: json['virtualCardImage'],
      applePayEnabled: json['applePayEnabled'] ?? false,
      googlePayEnabled: json['googlePayEnabled'] ?? false,
      createdAt: json['createdAt'] != null 
        ? DateTime.parse(json['createdAt']) 
        : DateTime.now(),
    );
  }

  String get maskedNumber => '•••• •••• •••• ${cardNumber.substring(cardNumber.length - 4)}';
}

class CardTransaction {
  final String id;
  final String cardId;
  final String userId;
  final double amount;
  final String currency;
  final String merchantName;
  final String merchantCategory;
  final String type;
  final String status;
  final String? receiptUrl;
  final DateTime timestamp;
  final String? location;

  CardTransaction({
    required this.id,
    required this.cardId,
    required this.userId,
    required this.amount,
    required this.currency,
    required this.merchantName,
    required this.merchantCategory,
    required this.type,
    required this.status,
    this.receiptUrl,
    required this.timestamp,
    this.location,
  });
}

class CardFundingSource {
  final String id;
  final String userId;
  final String token;
  final String symbol;
  final double balance;
  final double dailyLimit;
  final double monthlyLimit;
  final bool isDefault;

  CardFundingSource({
    required this.id,
    required this.userId,
    required this.token,
    required this.symbol,
    required this.balance,
    required this.dailyLimit,
    required this.monthlyLimit,
    required this.isDefault,
  });
}

class CardSettings {
  final String userId;
  final bool internationalEnabled;
  final bool onlinePaymentsEnabled;
  final bool contactlessEnabled;
  final double defaultDailyLimit;
  final double defaultMonthlyLimit;
  final List<String> blockedMerchants;

  CardSettings({
    required this.userId,
    required this.internationalEnabled,
    required this.onlinePaymentsEnabled,
    required this.contactlessEnabled,
    required this.defaultDailyLimit,
    required this.defaultMonthlyLimit,
    required this.blockedMerchants,
  });
}

class CardService {
  static final Map<String, List<CryptoCard>> _cards = {};
  static final Map<String, List<CardTransaction>> _transactions = {};
  static final Map<String, List<CardFundingSource>> _fundingSources = {};
  static final Map<String, CardSettings> _settings = {};

  static String _generateCardNumber() {
    final prefix = '4532';
    final random = DateTime.now().millisecondsSinceEpoch.toString();
    return '$prefix${random.substring(0, 12)}';
  }

  static String _generateCVV() {
    return (DateTime.now().millisecondsSinceEpoch % 1000).toString().padLeft(3, '0');
  }

  static String _generateExpiry() {
    final now = DateTime.now();
    final year = (now.year + 3).toString().substring(2);
    final month = (now.month + 1).toString().padLeft(2, '0');
    return '$month/$year';
  }

  static Future<List<CryptoCard>> getCards(String userId) async {
    if (_cards.containsKey(userId)) {
      return _cards[userId]!;
    }
    return [];
  }

  static Future<CryptoCard> createVirtualCard({
    required String userId,
    required String cardHolder,
    required String type,
    required String network,
    double dailyLimit = 10000,
    double monthlyLimit = 100000,
  }) async {
    final card = CryptoCard(
      id: 'card_${DateTime.now().millisecondsSinceEpoch}',
      userId: userId,
      cardNumber: _generateCardNumber(),
      cardHolder: cardHolder,
      expiryDate: _generateExpiry(),
      cvv: _generateCVV(),
      type: type,
      network: network,
      status: 'ACTIVE',
      dailyLimit: dailyLimit,
      monthlyLimit: monthlyLimit,
      dailySpent: 0,
      monthlySpent: 0,
      applePayEnabled: true,
      googlePayEnabled: true,
      createdAt: DateTime.now(),
    );
    
    _cards[userId] = [...(_cards[userId] ?? []), card];
    return card;
  }

  static Future<CryptoCard> createPhysicalCard({
    required String userId,
    required String cardHolder,
    required String shippingAddress,
    double dailyLimit = 10000,
    double monthlyLimit = 100000,
  }) async {
    final card = CryptoCard(
      id: 'card_physical_${DateTime.now().millisecondsSinceEpoch}',
      userId: userId,
      cardNumber: _generateCardNumber(),
      cardHolder: cardHolder,
      expiryDate: _generateExpiry(),
      cvv: _generateCVV(),
      type: 'PHYSICAL',
      network: 'VISA',
      status: 'PENDING_ACTIVATION',
      dailyLimit: dailyLimit,
      monthlyLimit: monthlyLimit,
      dailySpent: 0,
      monthlySpent: 0,
      applePayEnabled: true,
      googlePayEnabled: true,
      createdAt: DateTime.now(),
    );
    
    _cards[userId] = [...(_cards[userId] ?? []), card];
    return card;
  }

  static Future<void> freezeCard(String userId, String cardId) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: 'FROZEN',
        dailyLimit: card.dailyLimit,
        monthlyLimit: card.monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: card.applePayEnabled,
        googlePayEnabled: card.googlePayEnabled,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }

  static Future<void> unfreezeCard(String userId, String cardId) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: 'ACTIVE',
        dailyLimit: card.dailyLimit,
        monthlyLimit: card.monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: card.applePayEnabled,
        googlePayEnabled: card.googlePayEnabled,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }

  static Future<void> terminateCard(String userId, String cardId) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: 'TERMINATED',
        dailyLimit: card.dailyLimit,
        monthlyLimit: card.monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: false,
        googlePayEnabled: false,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }

  static Future<List<CardTransaction>> getTransactions(String userId, {String? cardId}) async {
    final allTxns = _transactions[userId] ?? [];
    if (cardId != null) {
      return allTxns.where((t) => t.cardId == cardId).toList();
    }
    return allTxns;
  }

  static Future<CardTransaction> processPayment({
    required String userId,
    required String cardId,
    required double amount,
    required String currency,
    required String merchantName,
    required String merchantCategory,
  }) async {
    final cards = _cards[userId] ?? [];
    final cardIndex = cards.indexWhere((c) => c.id == cardId);
    
    if (cardIndex == -1) {
      throw Exception('Card not found');
    }
    
    final card = cards[cardIndex];
    if (card.status != 'ACTIVE') {
      throw Exception('Card is not active');
    }
    
    if (card.dailySpent + amount > card.dailyLimit) {
      throw Exception('Daily limit exceeded');
    }
    
    if (card.monthlySpent + amount > card.monthlyLimit) {
      throw Exception('Monthly limit exceeded');
    }
    
    final transaction = CardTransaction(
      id: 'txn_${DateTime.now().millisecondsSinceEpoch}',
      cardId: cardId,
      userId: userId,
      amount: amount,
      currency: currency,
      merchantName: merchantName,
      merchantCategory: merchantCategory,
      type: 'PURCHASE',
      status: 'COMPLETED',
      timestamp: DateTime.now(),
    );
    
    _transactions[userId] = [...(_transactions[userId] ?? []), transaction];
    
    final updatedCard = CryptoCard(
      id: card.id,
      userId: card.userId,
      cardNumber: card.cardNumber,
      cardHolder: card.cardHolder,
      expiryDate: card.expiryDate,
      cvv: card.cvv,
      type: card.type,
      network: card.network,
      status: card.status,
      dailyLimit: card.dailyLimit,
      monthlyLimit: card.monthlyLimit,
      dailySpent: card.dailySpent + amount,
      monthlySpent: card.monthlySpent + amount,
      virtualCardImage: card.virtualCardImage,
      applePayEnabled: card.applePayEnabled,
      googlePayEnabled: card.googlePayEnabled,
      createdAt: card.createdAt,
    );
    cards[cardIndex] = updatedCard;
    _cards[userId] = cards;
    
    return transaction;
  }

  static Future<List<CardFundingSource>> getFundingSources(String userId) async {
    return _fundingSources[userId] ?? [];
  }

  static Future<void> addFundingSource({
    required String userId,
    required String token,
    required String symbol,
    required double balance,
    bool isDefault = false,
  }) async {
    final source = CardFundingSource(
      id: 'fund_${DateTime.now().millisecondsSinceEpoch}',
      userId: userId,
      token: token,
      symbol: symbol,
      balance: balance,
      dailyLimit: 10000,
      monthlyLimit: 100000,
      isDefault: isDefault,
    );
    
    final sources = _fundingSources[userId] ?? [];
    if (isDefault) {
      for (int i = 0; i < sources.length; i++) {
        sources[i] = CardFundingSource(
          id: sources[i].id,
          userId: sources[i].userId,
          token: sources[i].token,
          symbol: sources[i].symbol,
          balance: sources[i].balance,
          dailyLimit: sources[i].dailyLimit,
          monthlyLimit: sources[i].monthlyLimit,
          isDefault: false,
        );
      }
    }
    _fundingSources[userId] = [...sources, source];
  }

  static Future<CardSettings> getCardSettings(String userId) async {
    if (_settings.containsKey(userId)) {
      return _settings[userId]!;
    }
    
    final settings = CardSettings(
      userId: userId,
      internationalEnabled: true,
      onlinePaymentsEnabled: true,
      contactlessEnabled: true,
      defaultDailyLimit: 10000,
      defaultMonthlyLimit: 100000,
      blockedMerchants: [],
    );
    _settings[userId] = settings;
    return settings;
  }

  static Future<void> updateCardSettings(String userId, CardSettings settings) async {
    _settings[userId] = settings;
  }

  static Future<void> setCardLimit({
    required String userId,
    required String cardId,
    required double dailyLimit,
    required double monthlyLimit,
  }) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: card.status,
        dailyLimit: dailyLimit,
        monthlyLimit: monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: card.applePayEnabled,
        googlePayEnabled: card.googlePayEnabled,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }

  static Future<String> getVirtualCardImage(String cardId) async {
    return 'https://api.tigerwallet.com/cards/$cardId/image';
  }

  static Future<void> enableApplePay(String userId, String cardId) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: card.status,
        dailyLimit: card.dailyLimit,
        monthlyLimit: card.monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: true,
        googlePayEnabled: card.googlePayEnabled,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }

  static Future<void> enableGooglePay(String userId, String cardId) async {
    final cards = _cards[userId] ?? [];
    final index = cards.indexWhere((c) => c.id == cardId);
    if (index != -1) {
      final card = cards[index];
      cards[index] = CryptoCard(
        id: card.id,
        userId: card.userId,
        cardNumber: card.cardNumber,
        cardHolder: card.cardHolder,
        expiryDate: card.expiryDate,
        cvv: card.cvv,
        type: card.type,
        network: card.network,
        status: card.status,
        dailyLimit: card.dailyLimit,
        monthlyLimit: card.monthlyLimit,
        dailySpent: card.dailySpent,
        monthlySpent: card.monthlySpent,
        virtualCardImage: card.virtualCardImage,
        applePayEnabled: card.applePayEnabled,
        googlePayEnabled: true,
        createdAt: card.createdAt,
      );
      _cards[userId] = cards;
    }
  }
}
